Core/include/Acts/Vertexing/SeedVertexFinder.ipp

```diff
@@ -0,0 +1,656 @@
+          }
+        }
+      }
+
+      float delta2Zto = (middlez - nearz + 1) * zBinLength;
+      float angle2Zto = std::atan2(rFarDelta[!isMiddleLess], delta2Zto) -
+                        m_cfg.maxXYZdeviation;
+      int farz_to = m_cfg.numZSlices;
+      if (angle2Zto > 0) {
+        farz_to =
+            (int)(rFarDelta[!isMiddleLess] / std::tan(angle2Zto) / zBinLength) +
+            middlez + 1;
+        if (farz_to > m_cfg.numZSlices) {
+          farz_to = m_cfg.numZSlices;
+        } else {
+          if (farz_to < 0) {
+            continue;
+          }
+        }
+      }
+
+      for (int farz = farz_from; farz < farz_to; farz++) {
+        // loop over near phi slices
+        for (int nearphi = 0; nearphi < m_cfg.numPhiSlices; ++nearphi) {
+          // skip slices that are empty anyway
+          if (std::fmod(m_cfg.useFracPhiSlices * nearphi, 1.0) >=
+              m_cfg.useFracPhiSlices) {
+            continue;
+          }
+
+          // loop over some middle phi slices
+          for (int middlephi_h = nearphi - phistep;
+               middlephi_h <= nearphi + phistep; ++middlephi_h) {
+            int middlephi =
+                (middlephi_h + m_cfg.numPhiSlices) % m_cfg.numPhiSlices;
+            // loop over some far phi slices
+            for (int farphi_h = middlephi - phistep;
+                 farphi_h <= middlephi + phistep; ++farphi_h) {
+              int farphi = (farphi_h + m_cfg.numPhiSlices) % m_cfg.numPhiSlices;
+
+              // for all near spacepoints in this slice
+              for (const auto& near_sp :
+                   sorted_spacepoints.at(0).at(nearphi).at(nearz)) {
+                Acts::ActsScalar phiA = near_sp.second;
+                // for all middle spacepoints in this slice
+                for (const auto& middle_sp :
+                     sorted_spacepoints.at(1).at(middlephi).at(middlez)) {
+                  Acts::ActsScalar phiB = middle_sp.second;
+                  Acts::ActsScalar delta_phiAB =
+                      detail::difference_periodic(phiA, phiB, 2 * M_PI);
+                  if (std::abs(delta_phiAB) > m_cfg.maxPhideviation) {
+                    continue;
+                  }
+                  // for all far spacepoints in this slice
+                  for (const auto& far_sp :
+                       sorted_spacepoints.at(2).at(farphi).at(farz)) {
+                    Acts::ActsScalar phiC = far_sp.second;
+                    Acts::ActsScalar delta_phiBC =
+                        detail::difference_periodic(phiB, phiC, 2 * M_PI);
+                    if (std::abs(delta_phiBC) > m_cfg.maxPhideviation) {
+                      continue;
+                    }
+
+                    Acts::SeedVertexFinder<spacepoint_t>::Triplet tr(
+                        *near_sp.first, *middle_sp.first, *far_sp.first);
+
+                    if (isTripletValid(tr)) {
+                      triplets.push_back(tr);
+                    }
+                  }
```

Comment:
this look is kind of deep :fearful: I guess this is how the algorithm works... maybe lambdas could help out here but not sure. at the very least I would comment the `}` closing braces with what was iterated

Reply:
I don't think we can get rid of the nesting. 
Theoretically, we could define some simple lambda functions but each lambda would be used maybe twice, so it's a very little benefit for lower readability.
Also, some pieces of code are _almost_ the same but they have e.g. `+1` here and `-1` there; have `true` here and `false` there and so on. If we were to define such function to accommodate that, it would be as complex as the actual expression which defies the purpose.
Thus I think it's better to leave it as it is now.